tests/test_stability_funs.py

```diff
@@ -349,3 +351,403 @@ def test_magwell_print(capsys):
         + "\n"
     )
     assert out.out == corr_out
+
+
+@pytest.mark.unit
+def test_ballooning_geometry(tmpdir_factory):
+    """Test the geometry coefficients used for the adjoint-ballooning solver.
+
+    The same coefficients are used for local gyrokinetic solvers which would
+    be useful when we couple DESC with GX/GS2 etc.
+    Observation: The larger the force error, the worse the tests behave. For
+    example, HELIOTRON coefficients are hard to match
+    """
+    psi = 0.5  # Actually rho^2 (normalized)
+    alpha = 0.0
+    ntor = 2
+
+    eq0 = desc.examples.get("W7-X")
+    eq1 = desc.examples.get("precise_QA")
+
+    eq_list = [eq0, eq1]
+
+    for eq in eq_list:
+        eq_keys = ["iota", "iota_r", "a", "rho", "psi"]
+
+        grid = QuadratureGrid(eq.L_grid, eq.M_grid, eq.N_grid, NFP=eq.NFP)
+        data_eq = eq.compute(eq_keys, grid=grid)
+
+        fi = interp1d(
+            data_eq["rho"][grid.unique_rho_idx],
+            data_eq["iota"][grid.unique_rho_idx],
+        )
+        fs = interp1d(
+            data_eq["rho"][grid.unique_rho_idx],
+            data_eq["iota_r"][grid.unique_rho_idx],
+        )
+
+        iotas = fi(np.sqrt(psi))
+        shears = fs(np.sqrt(psi))
+
+        rho = np.sqrt(psi)
+        N = (8 * eq.M_grid * eq.N_grid) * ntor + 1
+        zeta = np.linspace(-ntor * np.pi, ntor * np.pi, N)
+
+        data_keys = [
+            "p_r",
+            "psi",
+            "psi_r",
+            "sqrt(g)_PEST",
+            "|grad(psi)|^2",
+            "grad(|B|)",
+            "grad(alpha)",
+            "grad(psi)",
+            "B",
+            "grad(|B|)",
+            "kappa",
+            "iota",
+            "lambda_t",
+            "lambda_z",
+            "lambda_tt",
+            "lambda_zz",
+            "lambda_tz",
+            "g^aa",
+            "g^ra",
+            "g^rr",
+            "cvdrift",
+            "cvdrift0",
+            "|B|",
+            "B^zeta",
+        ]
+
+        grid = Grid.create_meshgrid(
+            [rho, alpha, zeta],
+            coordinates="raz",
+            period=(np.inf, 2 * np.pi, np.inf),
+        )
+
+        data = eq.compute(data_keys, grid=grid)
+
+        psi_s = data_eq["psi"][-1]
+        sign_psi = psi_s / np.abs(psi_s)
+        sign_iota = iotas / np.abs(iotas)
+        # normalizations
+        Lref = data_eq["a"]
+        Bref = 2 * np.abs(psi) / Lref**2
+
+        modB = data["|B|"]
+        x = Lref * np.sqrt(psi)
+        shat = -x / iotas * shears / Lref
+
+        psi_r = data["psi_r"]
+
+        grad_psi = data["grad(psi)"]
+        grad_psi_sq = data["|grad(psi)|^2"]
+        grad_alpha = data["grad(alpha)"]
+
+        g_sup_rr = data["g^rr"]
+        g_sup_ra = data["g^ra"]
+        g_sup_aa = data["g^aa"]
+
+        modB = data["|B|"]
+
+        B_sup_zeta = data["B^zeta"]
+
+        gds2 = np.array(dot(grad_alpha, grad_alpha)) * Lref**2 * psi
+        gds2_alt = g_sup_aa * Lref**2 * psi
```

Comment:
like why would we expect these two arrays to be different, isnt grad alpha * grad alpha = g_sup_aa?

Reply:
so I dont understand why we compare them later

Reply:
This one is a good catch. I think I got rid of the old correct code. Basically, I calculate |grad_alpha| manually and then compare |grad(alpha)|**2 with g^aa. Corrected now!